src/viem/interfaces/Whoopi.v0.ts

```diff
@@ -0,0 +1,112 @@
+import { parseUnits } from 'viem/utils/';
+
+import { NETWORK_RESOLVERS } from '../../core';
+import {
+  DEPLOYMENT_WHOOPI_AVALANCHE_FUJI_TESTNET_V0,
+  DEPLOYMENT_WHOOPI_AVALANCHE_MAINNET_V0,
+  PaymentToken,
+  RenftContractType,
+  RenftContractVersions,
+} from '../../core';
+import { Executor, SDK, SDKInterface } from '../base';
+
+export default class WhoopiV0SDK<
+  ContractType extends RenftContractType,
+  ContractVersion extends RenftContractVersions[ContractType]
+> extends SDK<ContractType, ContractVersion> {
+  protected supportedDeployments = [
+    DEPLOYMENT_WHOOPI_AVALANCHE_FUJI_TESTNET_V0,
+    DEPLOYMENT_WHOOPI_AVALANCHE_MAINNET_V0,
+  ];
+
+  constructor(args: SDKInterface<ContractType, ContractVersion>) {
+    super(args);
+    super.validate(this.supportedDeployments);
+  }
+
+  async lend(
+    nftAddress: string,
+    tokenId: string[],
+    upfrontRentFees: string[],
+    revShareBeneficiaries: string[][],
+    revSharePortions: number[][],
+    maxRentDurations: number[],
+    paymentTokens: PaymentToken[],
+    allowedRenters?: string[][]
+  ): Promise<ReturnType<Executor>> {
+    const revShares = [];
+
+    for (let i = 0; i < revShareBeneficiaries.length; i++) {
+      revShares.push([revShareBeneficiaries[i], revSharePortions[i]]);
+    }
+
+    const allowRenters = [];
+
+    if (allowedRenters) {
+      for (let i = 0; i < allowedRenters.length; i++) {
+        allowRenters.push([allowedRenters[i]]);
+      }
+    } else {
+      for (let i = 0; i < tokenId.length; i++) {
+        // outer array is for the AllowedRenters struct,
+        // and the inner is for its contents: allowedRenters
+        // ! we need this layering because graphprotocol cannot
+        // ! generate types for 2d arrays. So you have to wrap
+        // ! the outer array into a struct.
+        allowRenters.push([[]]);
+      }
+    }
+
+    return this.exec('lend', [
+      [nftAddress, tokenId, Array(tokenId.length).fill('0')],
+      upfrontRentFees?.map((fee, i) => {
+        const scale = NETWORK_RESOLVERS[this.network][paymentTokens[i]].scale;
+        // @ts-ignore something about tsdx is messing this up
+        return parseUnits(fee, scale);
```

Comment:
what's `toWhoopiScaledAmount()` ?

Reply:
https://github.com/re-nft/sdk/blob/92f54e73e86ac30a5ec6837c6222d17086176fab/src/utils.ts#L241

Reply:
Right, so to elaborate. This function calls [`parseFixed()`](https://github.com/ethers-io/ethers.js/blob/v5.7/packages/bignumber/src.ts/fixednumber.ts#L70) depending on the `scale` setting of a token on a network. Thing is that our `ethers` SDK requires the `BigNumber` ponyfill whereas `viem` utilizes browser native `BigNumber`. I think [`parseUnits(()`](https://viem.sh/docs/utilities/parseUnits.html) does the same as `parseFixed()`?

Reply:
I recall now. `toWhoopiScaledAmount` and `fromWhoopiScaledAmount` I have created so that front devs (the ones that used to wrok on legacy v2) didn't confuse the units. There is little value in these functions, that is why I have TODO to remove them on there. People should use parseFixed / parseUnits directly, whatever is more appropriate in the context where it is being used.

Reply:
Double checked. `parseFixed()` returns a ethers BigNumber. The following are equivalent:

```ts
assert.equals(
  BigInt(parseFixed('420', 18).toHexString()), // ethers
  parseUnits('420', 18)                        // viem
);
```